k-distribution/k-tutorial/1_basic/04_disambiguation/README.md

```diff
@@ -316,12 +315,24 @@ modify the definition so that you will get the alternative parse.
 2. Modify your solution from Lesson 1.3, Exercise 2 so that unary negation should
 bind tighter than multiplication and division, which should bind tighter than
 addition and subtraction, and each binary operator should be left associative.
-Write these priority and associativity declarations both inline and explicitly.
+Write these priority and associativity declarations explicitly and then
+try to think whether they can be written inline.
 
 3. Write a simple grammar containing at least one ambiguity that cannot be
 resolved via priority or associativity, and then use the `prefer` attribute to
 resolve that ambiguity.
 
+4. Explain why the following grammar is not labeled ambiguous by the K parser when parsing `abb`, then fix it.
+```k
+module EXERCISE4
+
+syntax Expr ::= "a" Expr "b"
+              | "abb"
+              | "b"
+
+endmodule
+```
+
```

Comment:
It's not labeled ambiguous because we have a scanner that matches the longest token.
Not sure how you would fix this.

Reply:
Right, that's the point... I thought it could be useful adding it so that the reader learns about this.

For me, the solution is to use a sequence of tokens "a" and "b", not a token "abb". This does not *fix* the grammar, it just makes the parser realize the ambiguity.
```k
module EXERCISE2

  syntax Expr ::= "a" Expr "b"
              | "a" "b" "b"
              | "b"

endmodule
```

```
[Error] Inner Parser: Parsing ambiguity.
1: syntax Expr ::= "a" Expr "b"
    `a_b_EXERCISE2_Expr_Expr`(`b_EXERCISE2_Expr`(.KList))
2: syntax Expr ::= "a" "b" "b"
    `abb_EXERCISE2_Expr`(.KList)
	Source(<command line: -e>)
	Location(1,1,1,4)
```

Maybe I should change the phrasing to "make the parser realize the ambiguity"? Or we could move it out of the exercises and add a note somewhere, but I thought the exercise is kind of fun.

Reply:
.Actually I thought of the exercise while thinking about why my solution to Exercise 3 doesn't work, so this shouldn't require any preliminary knowledge of K.

Reply:
Yeah. You convinced me.

Reply:
Yeah, I'm happy to include this as a neat illustration of why the scanner-based parser works the way it does.